File: src/display.ts
/** Rendering of metadata for the terminal. */

import Table from "cli-table3";
import pc from "picocolors";

import type { FullMetadata, Metadata } from "./engine.js";

function asNumber(value: unknown): number | undefined {
  if (typeof value === "number" && Number.isFinite(value)) return value;
  if (typeof value === "string" && value.trim() !== "") {
    const n = Number(value);
    if (Number.isFinite(n)) return n;
  }
  return undefined;
}

function defined(value: unknown): boolean {
  return value !== undefined && value !== null && value !== "";
}

export function formatExposure(value: unknown): string {
  const seconds = asNumber(value);
  if (seconds === undefined) return String(value);
  if (seconds >= 1) return `${seconds}s`;
  return `1/${Math.round(1 / seconds)}s`;
}

export function formatSize(value: unknown): string {
  let size = asNumber(value);
  if (size === undefined) return String(value);
  if (size < 1024) return `${size} B`;
  for (const unit of ["KB", "MB", "GB", "TB"]) {
    size /= 1024;
    if (size < 1024 || unit === "TB") return `${size.toFixed(1)} ${unit}`;
  }
  return String(value);
}

export function formatDuration(value: unknown): string {
  const total = asNumber(value);
  if (total === undefined) return String(value);
  const rounded = Math.round(total);
  const hours = Math.floor(rounded / 3600);
  const minutes = Math.floor((rounded % 3600) / 60);
  const seconds = rounded % 60;
  const pad = (n: number) => String(n).padStart(2, "0");
  if (hours) return `${hours}h ${pad(minutes)}m ${pad(seconds)}s`;
  return `${minutes}m ${pad(seconds)}s`;
}

export function formatGps(metadata: Metadata): string | undefined {
  const lat = asNumber(metadata.GPSLatitude);
  const lon = asNumber(metadata.GPSLongitude);
  if (lat === undefined || lon === undefined) {
    const coords = metadata.GPSCoordinates; // QuickTime videos
    return coords ? String(coords) : undefined;
  }
  let text = `${lat.toFixed(6)}, ${lon.toFixed(6)}`;
  const altitude = asNumber(metadata.GPSAltitude);
  if (altitude !== undefined) text += `  (altitude ${altitude.toFixed(1)}m)`;
  return text;
}

const GPS_KEYS = [
  "GPSLatitude", "GPSLongitude", "GPSAltitude",
  "GPSLatitudeRef", "GPSLongitudeRef", "GPSAltitudeRef",
  "GPSPosition", "GPSCoordinates",
];

// Shutter actuation counters vary by manufacturer (Nikon/Sony/Pentax use
// ShutterCount; others expose ImageCount or ImageNumber).
const SHUTTER_COUNT_KEYS = [
  "ShutterCount", "MechanicalShutterCount", "ImageCount", "ImageNumber",
];

const SERIAL_KEYS = ["SerialNumber", "InternalSerialNumber", "CameraSerialNumber"];

export interface ReportRows {
  /** Curated, photographer-relevant rows, most important first. */
  main: [string, string][];
  /** Every remaining tag, alphabetical. */
  other: [string, string][];
}

/** Build the full report: curated rows first, all remaining tags after. */
export function fullReportRows(full: FullMetadata): ReportRows {
  const { pretty: p, numeric: n } = full;
  const used = new Set<string>(["SourceFile"]);
  const main: [string, string][] = [];

  const add = (label: string, value: unknown, keys: string[]) => {
    for (const k of keys) used.add(k);
    if (defined(value)) main.push([label, String(value)]);
  };
  const firstPretty = (keys: string[]) => keys.map((k) => p[k]).find(defined);

  // File
  add("File", p.FileName, ["FileName"]);
  add("Folder", p.Directory, ["Directory"]);
  add("Type", p.FileType, ["FileType"]);
  add("Size", defined(n.FileSize) ? formatSize(n.FileSize) : undefined, ["FileSize"]);
  const width = asNumber(n.ImageWidth);
  const height = asNumber(n.ImageHeight);
  add(
    "Dimensions",
    width && height ? `${width} x ${height}` : undefined,
    ["ImageWidth", "ImageHeight", "ImageSize"],
  );
  add("Megapixels", p.Megapixels, ["Megapixels"]);

  // Camera & lens
  const camera = [p.Make, p.Model].filter(Boolean).join(" ");
  add("Camera", camera || undefined, ["Make", "Model"]);
  add("Serial number", firstPretty(SERIAL_KEYS), SERIAL_KEYS);
  const shutterCount = SHUTTER_COUNT_KEYS.map((k) => asNumber(n[k])).find(
    (v) => v !== undefined && v > 0,
  );
  add("Shutter count", shutterCount?.toLocaleString("en-US"), SHUTTER_COUNT_KEYS);
  add("Lens", firstPretty(["LensModel", "LensID", "Lens"]), [
    "LensModel", "LensID", "Lens",
  ]);

  // Exposure
  add("ISO", p.ISO, ["ISO"]);
  const fNumber = asNumber(n.FNumber);
  add("Aperture", fNumber !== undefined ? `f/${fNumber}` : undefined, ["FNumber"]);
  add(
    "Shutter",
    defined(n.ExposureTime) ? formatExposure(n.ExposureTime) : undefined,
    ["ExposureTime"],
  );
  const focal = asNumber(n.FocalLength);
  const focal35 = asNumber(n.FocalLengthIn35mmFormat);
  let focalText: string | undefined;
  if (focal !== undefined) {
    focalText = `${focal}mm`;
    if (focal35 !== undefined && focal35 !== focal) {
      focalText += `  (${focal35}mm in 35mm equiv.)`;
    }
  }
  add("Focal length", focalText, ["FocalLength", "FocalLengthIn35mmFormat"]);
  const ec = p.ExposureCompensation;
  add("Exposure comp.", defined(ec) && String(ec) !== "0" ? `${ec} EV` : undefined, [
    "ExposureCompensation",
  ]);
  add("Exposure program", p.ExposureProgram, ["ExposureProgram"]);
  add("Metering", p.MeteringMode, ["MeteringMode"]);
  add("Flash", p.Flash, ["Flash"]);
  add("White balance", p.WhiteBalance, ["WhiteBalance"]);
  add("Color space", p.ColorSpace, ["ColorSpace"]);
  add("Orientation", p.Orientation, ["Orientation"]);

  // Authorship
  add("Artist", p.Artist, ["Artist"]);
  add("Copyright", p.Copyright, ["Copyright"]);
  add("Rating", p.Rating, ["Rating"]);
  add("Software", p.Software, ["Software"]);

  // Dates
  add("Taken (DateTimeOriginal)", p.DateTimeOriginal, ["DateTimeOriginal"]);
  add("Created (CreateDate)", p.CreateDate, ["CreateDate"]);
  add("Modified (ModifyDate)", p.ModifyDate, ["ModifyDate"]);
  add("Timezone (OffsetTimeOriginal)", p.OffsetTimeOriginal, ["OffsetTimeOriginal"]);
  add("File modified", p.FileModifyDate, ["FileModifyDate"]);

  // Video
  add(
    "Duration",
    defined(n.Duration) ? formatDuration(n.Duration) : undefined,
    ["Duration"],
  );
  const fps = asNumber(n.VideoFrameRate);
  add("Frame rate", fps !== undefined ? `${fps} fps` : undefined, ["VideoFrameRate"]);
  add("Audio", p.AudioFormat, ["AudioFormat"]);

  // GPS (decimal degrees, ready to paste into a maps app)
  add("GPS", formatGps(n) ?? "— none —", GPS_KEYS);

  const other: [string, string][] = Object.keys(p)
    .filter((k) => !used.has(k))
    .sort()
    .map((k) => [k, String(p[k])]);

  return { main, other };
}

function newTable(head?: string[]): Table.Table {
  return new Table({
    head,
    style: { head: [], border: [] },
    wordWrap: true,
    colWidths: [32, 60],
  });
}

export function printFullReport(full: FullMetadata): void {
  const { main, other } = fullReportRows(full);
  console.log(pc.bold(pc.cyan(String(full.pretty.FileName ?? ""))));
  const table = newTable();
  for (const [label, value] of main) table.push([pc.bold(label), value]);
  console.log(table.toString());

  if (other.length > 0) {
    console.log(pc.dim(`All other tags (${other.length}):`));
    const rest = newTable();
    for (const [key, value] of other) rest.push([pc.dim(key), pc.dim(value)]);
    console.log(rest.toString());
  }
}

export function printAllTags(metadata: Metadata): void {
  console.log(pc.bold(pc.cyan(String(metadata.FileName ?? ""))));
  const table = newTable([pc.bold("Tag"), pc.bold("Value")]);
  for (const key of Object.keys(metadata).sort()) {
    if (key === "SourceFile") continue;
    table.push([key, String(metadata[key])]);
  }
  console.log(table.toString());
}

export function printSuccess(message: string): void {
  console.log(`${pc.green(pc.bold("✓"))} ${message}`);
}

export function printError(message: string): void {
  console.error(pc.red(message));
}

export function describeFiles(paths: string[]): string {
  if (paths.length === 1) {
    const p = paths[0];
    return p.split(/[/\\]/).pop() ?? p;
  }
  return `${paths.length} files`;
}

File: src/exifkit/display.py
"""Rendering of metadata for the terminal using rich."""

from __future__ import annotations

from fractions import Fraction
from pathlib import Path

from rich.console import Console
from rich.table import Table

console = Console()
error_console = Console(stderr=True, style="bold red")


def _format_exposure(value: object) -> str:
    try:
        seconds = float(value)  # type: ignore[arg-type]
    except (TypeError, ValueError):
        return str(value)
    if seconds >= 1:
        return f"{seconds:g}s"
    fraction = Fraction(seconds).limit_denominator(8000)
    return f"1/{round(fraction.denominator / fraction.numerator)}s"


def _format_gps(metadata: dict) -> str | None:
    lat = metadata.get("GPSLatitude")
    lon = metadata.get("GPSLongitude")
    if lat is None or lon is None:
        coords = metadata.get("GPSCoordinates")  # QuickTime videos
        return str(coords) if coords else None
    text = f"{float(lat):.6f}, {float(lon):.6f}"
    altitude = metadata.get("GPSAltitude")
    if altitude is not None:
        text += f"  (altitude {float(altitude):.1f}m)"
    return text


def _format_size(value: object) -> str:
    try:
        size = float(value)  # type: ignore[arg-type]
    except (TypeError, ValueError):
        return str(value)
    for unit in ("B", "KB", "MB", "GB", "TB"):
        if size < 1024 or unit == "TB":
            return f"{size:.1f} {unit}" if unit != "B" else f"{int(size)} B"
        size /= 1024
    return str(value)


def _format_duration(value: object) -> str:
    try:
        total = round(float(value))  # type: ignore[arg-type]
    except (TypeError, ValueError):
        return str(value)
    minutes, seconds = divmod(total, 60)
    hours, minutes = divmod(minutes, 60)
    if hours:
        return f"{hours}h {minutes:02d}m {seconds:02d}s"
    return f"{minutes}m {seconds:02d}s"


def summary_rows(metadata: dict) -> list[tuple[str, str]]:
    """Build (label, value) rows for the curated summary view."""
    rows: list[tuple[str, str]] = []

    def add(label: str, value: object | None) -> None:
        if value not in (None, "", 0):
            rows.append((label, str(value)))

    add("File", metadata.get("FileName"))
    add("Type", metadata.get("FileType"))
    if metadata.get("FileSize") is not None:
        add("Size", _format_size(metadata["FileSize"]))
    width, height = metadata.get("ImageWidth"), metadata.get("ImageHeight")
    if width and height:
        add("Dimensions", f"{width} x {height}")

    make, model = metadata.get("Make"), metadata.get("Model")
    camera = " ".join(str(p) for p in (make, model) if p)
    add("Camera", camera or None)
    add("Lens", metadata.get("LensModel") or metadata.get("LensID"))

    if metadata.get("ISO") is not None:
        add("ISO", metadata.get("ISO"))
    if metadata.get("FNumber") is not None:
        add("Aperture", f"f/{metadata['FNumber']:g}")
    if metadata.get("ExposureTime") is not None:
        add("Shutter", _format_exposure(metadata["ExposureTime"]))
    if metadata.get("FocalLength") is not None:
        add("Focal length", f"{metadata['FocalLength']:g}mm")

    add("Taken (DateTimeOriginal)", metadata.get("DateTimeOriginal"))
    add("Created (CreateDate)", metadata.get("CreateDate"))
    add("Modified (ModifyDate)", metadata.get("ModifyDate"))
    add("File modified", metadata.get("FileModifyDate"))

    if metadata.get("Duration") is not None:
        add("Duration", _format_duration(metadata["Duration"]))

    gps = _format_gps(metadata)
    add("GPS", gps or "— none —")
    return rows


def print_summary(metadata: dict) -> None:
    table = Table(
        title=metadata.get("FileName", ""),
        show_header=False,
        title_style="bold cyan",
        border_style="dim",
    )
    table.add_column("Field", style="bold", no_wrap=True)
    table.add_column("Value")
    for label, value in summary_rows(metadata):
        table.add_row(label, value)
    console.print(table)


def print_all_tags(metadata: dict) -> None:
    table = Table(
        title=metadata.get("FileName", ""),
        title_style="bold cyan",
        border_style="dim",
    )
    table.add_column("Tag", style="bold", no_wrap=True)
    table.add_column("Value", overflow="fold")
    for key in sorted(metadata):
        if key == "SourceFile":
            continue
        table.add_row(key, str(metadata[key]))
    console.print(table)


def print_success(message: str) -> None:
    console.print(f"[bold green]✓[/bold green] {message}")


def print_error(message: str) -> None:
    error_console.print(message)


def describe_files(paths: list[Path]) -> str:
    if len(paths) == 1:
        return paths[0].name
    return f"{len(paths)} files"
